core/src/main/scala/caliban/GraphQL.scala

```diff
@@ -89,16 +91,16 @@ trait GraphQL[-R] { self =>
           queryExecution: QueryExecution
         ): URIO[R, GraphQLResponse[CalibanError]] =
           decompose(wrappers).flatMap {
-            case (overallWrappers, parsingWrappers, validationWrappers, executionWrappers, fieldWrappers) =>
+            case (overallWrappers, parsingWrappers, validationWrappers, executionWrappers, fieldWrappers, _) =>
               wrap((request: GraphQLRequest) =>
                 (for {
                   doc   <- wrap(Parser.parseQuery)(parsingWrappers, request.query.getOrElse(""))
                   intro = Introspector.isIntrospection(doc)
                   _ <- IO.when(intro && !enableIntrospection) {
                         IO.fail(CalibanError.ValidationError("Introspection is disabled", ""))
                       }
-                  typeToValidate  = if (intro) introspectionRootType else rootType
-                  schemaToExecute = if (intro) introspectionRootSchema else schema
+                  typeToValidate  <- if (intro) introspectionRootType else ZIO.succeed(rootType)
```

Comment:
We don't need to re-create `introspectionRootType` and `introExecSchema` at every request, it can be done in `def interpreter` in the outer effect.

Reply:
And if you do this, there is no need for splitting `introCheckSchema` and `introExecSchema` I think.

Reply:
this part look really ugly agreed, I have to do this because of signature of check method: override def check(query: String): IO[CalibanError, Unit], meaning check and execute can't use same effect, as I need R at execute. I'm open for suggestions. check won't be able to run Wrapper[R]

Reply:
I will make a work-around. in case of ::check I will pass Any as env, and Nil for wrappers. this will allow me avoid splitting